File: ci/eval-packages.ts
import { realpathSync } from "node:fs";
import {
	isRecord,
	parseJson,
	prettyJson,
	requireRecord,
	requireString,
	run,
	writeTextFile,
} from "./lib.ts";

const APPLY_EXPRESSION =
	"packages: builtins.mapAttrs (_: package: { " +
	"path = package.outPath; " +
	"version = package.version or null; " +
	"}) packages";

export type PackageOutput = Readonly<{
	path: string;
	version: string | null;
}>;

export type PackageSet = Readonly<Record<string, PackageOutput>>;

export type ChangedPackage = Readonly<{
	after: PackageOutput;
	before: PackageOutput;
	name: string;
}>;

export type EvalComparison = Readonly<{
	added: readonly string[];
	changed: readonly ChangedPackage[];
	mergedCount: number;
	removed: readonly string[];
	system: string;
	targetCount: number;
	unchangedCount: number;
}>;

type Arguments = Readonly<{
	merged: string;
	output: string;
	system: string;
	target: string;
}>;

type SnapshotArguments = Readonly<{
	output: string;
	repository: string;
	system: string;
}>;

function parseCount(value: unknown, name: string): number {
	if (!Number.isInteger(value) || typeof value !== "number" || value < 0) {
		throw new Error(`${name} must be a non-negative integer`);
	}
	return value;
}

function parseNames(value: unknown, name: string): readonly string[] {
	if (!Array.isArray(value)) {
		throw new Error(`${name} must be an array`);
	}
	return value.map((item, index) => requireString(item, `${name}[${index}]`));
}

function parsePackageOutput(value: unknown, name: string): PackageOutput {
	const output = requireRecord(value, name);
	const version = output.version;
	if (version !== null && typeof version !== "string") {
		throw new Error(`${name}.version must be a string or null`);
	}
	return {
		path: requireString(output.path, `${name}.path`),
		version,
	};
}

export function parseEvalComparison(value: unknown): EvalComparison {
	const comparison = requireRecord(value, "eval comparison");
	if (!Array.isArray(comparison.changed)) {
		throw new Error("eval comparison.changed must be an array");
	}
	const changed = comparison.changed.map((item, index) => {
		const change = requireRecord(item, `eval comparison.changed[${index}]`);
		return {
			after: parsePackageOutput(
				change.after,
				`eval comparison.changed[${index}].after`,
			),
			before: parsePackageOutput(
				change.before,
				`eval comparison.changed[${index}].before`,
			),
			name: requireString(
				change.name,
				`eval comparison.changed[${index}].name`,
			),
		} satisfies ChangedPackage;
	});
	return {
		added: parseNames(comparison.added, "eval comparison.added"),
		changed,
		mergedCount: parseCount(
			comparison.mergedCount,
			"eval comparison.mergedCount",
		),
		removed: parseNames(comparison.removed, "eval comparison.removed"),
		system: requireString(comparison.system, "eval comparison.system"),
		targetCount: parseCount(
			comparison.targetCount,
			"eval comparison.targetCount",
		),
		unchangedCount: parseCount(
			comparison.unchangedCount,
			"eval comparison.unchangedCount",
		),
	};
}

export function parsePackageSet(value: unknown, system: string): PackageSet {
	if (!isRecord(value)) {
		throw new Error(`Package evaluation for ${system} returned a non-object`);
	}
	const packages: Record<string, PackageOutput> = {};
	for (const [name, item] of Object.entries(value)) {
		if (!isRecord(item) || typeof item.path !== "string") {
			throw new Error(`Package ${name} has no output path on ${system}`);
		}
		if (item.version !== null && typeof item.version !== "string") {
			throw new Error(`Package ${name} has an invalid version on ${system}`);
		}
		packages[name] = { path: item.path, version: item.version };
	}
	return packages;
}

export async function evaluatePackages(
	repository: string,
	system: string,
): Promise<PackageSet> {
	const absoluteRepository = realpathSync(repository);
	const result = await run(
		[
			"nix",
			"eval",
			"--json",
			"--no-write-lock-file",
			"--option",
			"accept-flake-config",
			"false",
			"--option",
			"allow-import-from-derivation",
			"false",
			`path:${absoluteRepository}#packages.${system}`,
			"--apply",
			APPLY_EXPRESSION,
		],
		{ capture: true },
	);
	return parsePackageSet(
		parseJson(result.stdout, `package evaluation for ${system}`),
		system,
	);
}

export function comparePackages(
	target: PackageSet,
	merged: PackageSet,
	system: string,
): EvalComparison {
	const targetNames = new Set(Object.keys(target));
	const mergedNames = new Set(Object.keys(merged));
	const commonNames = [...targetNames]
		.filter((name) => mergedNames.has(name))
		.sort();
	const changed = commonNames.flatMap((name) => {
		const before = target[name];
		const after = merged[name];
		if (
			before === undefined ||
			after === undefined ||
			before.path === after.path
		) {
			return [];
		}
		return [{ after, before, name }];
	});
	return {
		added: [...mergedNames].filter((name) => !targetNames.has(name)).sort(),
		changed,
		mergedCount: mergedNames.size,
		removed: [...targetNames].filter((name) => !mergedNames.has(name)).sort(),
		system,
		targetCount: targetNames.size,
		unchangedCount: commonNames.length - changed.length,
	};
}

export function markdownSummary(result: EvalComparison): string {
	const lines = [
		`## Eval: ${result.system}`,
		"",
		"| Result | Count |",
		"| --- | ---: |",
		`| Added | ${result.added.length} |`,
		`| Removed | ${result.removed.length} |`,
		`| Changed | ${result.changed.length} |`,
		`| Unchanged | ${result.unchangedCount} |`,
	];
	const sections: readonly Readonly<{
		names: readonly string[];
		title: string;
	}>[] = [
		{ names: result.added, title: "Added packages" },
		{ names: result.removed, title: "Removed packages" },
		{
			names: result.changed.map(({ name }) => name),
			title: "Changed packages",
		},
	];
	for (const { names, title } of sections) {
		if (names.length > 0) {
			lines.push(
				"",
				`### ${title}`,
				"",
				...names.map((name) => `- \`${name}\``),
			);
		}
	}
	return `${lines.join("\n")}\n`;
}

export function parseArguments(args: readonly string[]): Arguments {
	const values = new Map<string, string>();
	for (let index = 0; index < args.length; index += 2) {
		const flag = args[index];
		const value = args[index + 1];
		if (!flag?.startsWith("--") || value === undefined) {
			throw new Error(
				"Expected --target, --merged, --system and --output arguments",
			);
		}
		values.set(flag.slice(2), value);
	}
	const target = values.get("target");
	const merged = values.get("merged");
	const system = values.get("system");
	const output = values.get("output");
	if (!target || !merged || !system || !output || values.size !== 4) {
		throw new Error(
			"Expected --target, --merged, --system and --output arguments",
		);
	}
	return { merged, output, system, target };
}

function parseSnapshotArguments(args: readonly string[]): SnapshotArguments {
	const values = new Map<string, string>();
	for (let index = 0; index < args.length; index += 2) {
		const flag = args[index];
		const value = args[index + 1];
		if (!flag?.startsWith("--") || value === undefined) {
			throw new Error("Expected --repository, --system and --output arguments");
		}
		values.set(flag.slice(2), value);
	}
	const repository = values.get("repository");
	const system = values.get("system");
	const output = values.get("output");
	if (!repository || !system || !output || values.size !== 3) {
		throw new Error("Expected --repository, --system and --output arguments");
	}
	return { output, repository, system };
}

export async function evalSnapshot(args: readonly string[]): Promise<void> {
	const { output, repository, system } = parseSnapshotArguments(args);
	writeTextFile(output, prettyJson(await evaluatePackages(repository, system)));
}

export async function evalPackages(args: readonly string[]): Promise<void> {
	const { merged, output, system, target } = parseArguments(args);
	const comparison = comparePackages(
		await evaluatePackages(target, system),
		await evaluatePackages(merged, system),
		system,
	);
	writeTextFile(output, prettyJson(comparison));
	console.log(markdownSummary(comparison));
}

File: ci/eval_packages.py
"""Evaluate and compare package outputs for two flake checkouts."""

from __future__ import annotations

import argparse
import json
import os
import subprocess
from pathlib import Path
from typing import Optional, cast

Package = dict[str, Optional[str]]
PackageSet = dict[str, Package]

APPLY_EXPRESSION = (
    "packages: builtins.mapAttrs (_: package: { "
    "path = package.outPath; "
    "version = package.version or null; "
    "}) packages"
)


def evaluate_packages(repository: Path, system: str) -> PackageSet:
    command = [
        "nix",
        "eval",
        "--json",
        "--no-write-lock-file",
        "--option",
        "accept-flake-config",
        "false",
        "--option",
        "allow-import-from-derivation",
        "false",
        f"path:{repository.resolve()}#packages.{system}",
        "--apply",
        APPLY_EXPRESSION,
    ]
    result = subprocess.run(command, check=True, capture_output=True, text=True)
    payload: object = json.loads(result.stdout)
    if not isinstance(payload, dict):
        raise RuntimeError(f"Package evaluation for {system} returned a non-object")

    packages = cast(dict[object, object], payload)
    validated: PackageSet = {}
    for name, value in packages.items():
        if not isinstance(name, str) or not isinstance(value, dict):
            raise RuntimeError(f"Package evaluation for {system} returned invalid data")
        path = value.get("path")
        version = value.get("version")
        if not isinstance(path, str):
            raise RuntimeError(f"Package {name} has no output path on {system}")
        if version is not None and not isinstance(version, str):
            raise RuntimeError(f"Package {name} has an invalid version on {system}")
        validated[name] = {"path": path, "version": version}
    return validated


def comparison(
    target: PackageSet, merged: PackageSet, system: str
) -> dict[str, object]:
    target_names = set(target)
    merged_names = set(merged)
    common_names = target_names & merged_names
    changed = [
        {
            "after": merged[name],
            "before": target[name],
            "name": name,
        }
        for name in sorted(common_names)
        if target[name]["path"] != merged[name]["path"]
    ]
    return {
        "added": sorted(merged_names - target_names),
        "changed": changed,
        "mergedCount": len(merged),
        "removed": sorted(target_names - merged_names),
        "system": system,
        "targetCount": len(target),
        "unchangedCount": len(common_names) - len(changed),
    }


def markdown_summary(result: dict[str, object]) -> str:
    added = cast(list[str], result["added"])
    changed = cast(list[dict[str, object]], result["changed"])
    removed = cast(list[str], result["removed"])
    lines = [
        f"## Eval: {result['system']}",
        "",
        "| Result | Count |",
        "| --- | ---: |",
        f"| Added | {len(added)} |",
        f"| Removed | {len(removed)} |",
        f"| Changed | {len(changed)} |",
        f"| Unchanged | {result['unchangedCount']} |",
    ]
    sections = (
        ("Added packages", added),
        ("Removed packages", removed),
        ("Changed packages", [cast(str, item["name"]) for item in changed]),
    )
    for title, names in sections:
        if names:
            lines.extend(["", f"### {title}", "", *[f"- `{name}`" for name in names]])
    return "\n".join(lines) + "\n"


def arguments() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--merged", required=True, type=Path)
    parser.add_argument("--output", required=True, type=Path)
    parser.add_argument("--system", required=True)
    parser.add_argument("--target", required=True, type=Path)
    return parser.parse_args()


def main() -> None:
    args = arguments()
    target = evaluate_packages(args.target, args.system)
    merged = evaluate_packages(args.merged, args.system)
    result = comparison(target, merged, args.system)
    args.output.write_text(json.dumps(result, indent=2, sort_keys=True) + "\n")

    summary = markdown_summary(result)
    print(summary)
    if summary_path := os.environ.get("GITHUB_STEP_SUMMARY"):
        with Path(summary_path).open("a") as file:
            file.write(summary)


if __name__ == "__main__":
    main()
